src/group/managed_group/mod.rs

```diff
@@ -0,0 +1,276 @@
+/// those messages are not automatically appended to this queue, instead they have be processed again through `process_message()`.
+/// This allows the Delivery Service to reject them (e.g. if they reference the wrong epoch).
+///
+/// If incoming messages or applied operations are semantically or syntactily incorrect, the function will return a corresponding
+/// error message and the state of the group will remain unchanged.
+///
+/// The application policy for the group can be enforced by implementing the validator callback functions and selectively allowing/
+/// disallowing each operation (see `ManagedGroupCallbacks`)
+///
+/// Changes to the group state are dispatched as events through callback functions (see ManagedGroupCallbacks).
+pub struct ManagedGroup {
+    managed_group_config: ManagedGroupConfig,
+    group: MlsGroup,
+    proposal_queue: ProposalQueue,
+    own_kpbs: Vec<KeyPackageBundle>,
+}
+
+impl ManagedGroup {
+    // === Group creation ===
+
+    /// Creates a new group from scratch with only the creator as a member.
+    pub fn new(
+        credential_bundle: &CredentialBundle,
+        managed_group_config: &ManagedGroupConfig,
+        group_id: GroupId,
+        ciphersuite_name: CiphersuiteName,
+        key_package_bundle: KeyPackageBundle,
+    ) -> Result<Self, GroupError> {
+        let group = MlsGroup::new(
+            &group_id.as_slice(),
+            ciphersuite_name,
+            key_package_bundle,
+            GroupConfig::default(),
+        )?;
+
+        Ok(ManagedGroup {
+            managed_group_config: managed_group_config.clone(),
+            group,
+            proposal_queue: ProposalQueue::new(),
+            own_kpbs: vec![],
+        })
+    }
+
+    /// Creates a new group from a `Welcome` message
+    pub fn new_from_welcome(
+        credential_bundle: &CredentialBundle,
+        managed_group_config: &ManagedGroupConfig,
+        welcome: Welcome,
+        ratchet_tree: Option<Vec<Option<Node>>>,
+        key_package_bundle: KeyPackageBundle,
+    ) -> Result<Self, WelcomeError> {
+        let group = MlsGroup::new_from_welcome(welcome, ratchet_tree, key_package_bundle)?;
+        Ok(ManagedGroup {
+            managed_group_config: managed_group_config.clone(),
+            group,
+            proposal_queue: ProposalQueue::new(),
+            own_kpbs: vec![],
+        })
+    }
+
+    /// Creates a new group with the creator and additional members
+    pub fn new_with_members(
+        credential_bundle: &CredentialBundle,
+        managed_group_config: &ManagedGroupConfig,
+        group_id: GroupId,
+        ciphersuite_name: CiphersuiteName,
+        key_package_bundle: KeyPackageBundle,
+        members: Vec<KeyPackage>,
+    ) -> Result<Self, GroupError> {
+        unimplemented!()
+    }
+
+    // === Membership management ===
+
+    /// Adds members to the group
+    pub fn add_members(
+        &mut self,
+        credential_bundle: &CredentialBundle,
+        aad: &[u8],
```

Comment:
Would it make sense to set `AAD` for the group to re-use it? It's nowhere specified how this is handled. But I'd probably build the `aad` from some group aad and maybe other information such as sender etc. This could make the `aad` here optional and such simpler. (It's already optional as it can be empty, but that's not clear right now.)

Reply:
Yes, that's an open question and I had the same idea. Right now you can only specify it for application messages, and you can also just read it when application messages are received. HS messages can also have an AAD, but it's being ignored.
One option would indeed to set it globally and modify it when needed (should it be persisted?). In addition, it also needs to be returned when HS messages are parsed, which is much trickier and doesn't fit into the API right now, unless we make `process_message()` return the AAD.

Reply:
Ok, I fixed this now I think.